draft-irtf-cfrg-opaque.md

```diff
@@ -398,145 +398,331 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
+2. Output (nil, pkc)
 ~~~
 
-The `custom_identifier` mode defines `SecretCredentials` and `CleartextCredentials` as follows:
+OPAQUE-3DH uses HashToScalar(seed) to derive the client's secret key in this mode.
+~~~
+DeriveKeyPair(prk)
 
+Input:
+- prk, pseudo-random key used as seed
+
+Output:
+- sk, the client's private key
+- pk, the client's public key
+
+Steps:
+1. skc = HashToScalar(prk, "OPAQUE-HashToScalar")
+2. pck = skc * G
+3. Output (skc, pkc)
 ~~~
-struct {
-  opaque client_private_key[Nsk];
-} SecretCredentials;
 
-struct {
-  opaque server_public_key[Npk];
-  opaque client_identity<0..2^16-1>;
-  opaque server_identity<0..2^16-1>;
-} CleartextCredentials;
 ~~~
+RecoverSecret(prk, inner_env)
+
+Input:
+- prk, pseudo-random key
+- inner_env, an InnerEnvelope structure. Not used in this function, it only serves to comply with the API
+
+Output:
+- secret_creds, SecretCredentials
+
+Steps:
+1. skc = KeyGenFromSeed(prk)
+2. Create SecretCredentials secret_creds with skc
+3. Output secret_creds
+~~~
+
+### External Key mode {#external-mode}
 
-These credentials are embedded into the following `Envelope` structure with
-encryption and authentication.
+In this mode, the application provides the OPAQUE instantiation with the client's authentication material through the `Credentials` structure. An encryption key is generated from the hardened OPRF output and used to encrypt the client's private key, which is then stored encrypted in the `Envelope`. This allows for keys of any kind as long as they are compatible with the AKE mechanism (e.g. RSA or ECDSA in the SIGMA-I setup, not specified in this document). Applications can therefore use existing key pairs for OPAQUE-3DH.
 
 ~~~
 struct {
-  EnvelopeMode mode;
-  opaque nonce[32];
-  opaque encrypted_creds[Nsk];
+ opaque encrypted_creds[Nsk];
 } InnerEnvelope;
+~~~
+
+encrypted_creds : Encoding of encrypted `SecretCredentials`. Authentication is ensured with the `AuthTag` in the envelop that covers this `InnerEnvelope`.
+
+Functions: Envelope and Recover
+~~~
+BuildInnerEnvelope(prk, creds)
+
+Input:
+- prk, pseudo-random key
+- creds, Credentials structure
+
+Output:
+- inner_env, an InnerEnvelope structure
+- client_public_key, the client's AKE public key
+
+Steps:
+1. Create SecretCredentials secret_creds with creds.client_private_key
+2. pseudorandom_pad = Expand(prk, "Pad", len(secret_creds))
+3. encrypted_creds = xor(secret_creds, pseudorandom_pad)
+4. Create InnerEnvelope inner_env with encrypted_creds
+5. Output (inner_env, creds.client_public_key)
+~~~
+
+~~~
+RecoverSecret(prk, inner_env)
+
+Input:
+- prk, pseudo-random key
+- inner_env, an InnerEnvelope structure
+
+Output:
+- secret_creds, SecretCredentials
+
+Steps:
+1. encrypted_creds = inner_env.encrypted_creds
+2. pseudorandom_pad = Expand(prk, "Pad", len(encrypted_creds))
+3. secret_creds = xor(encrypted_creds, pseudorandom_pad)
+4. Output secret_creds
+~~~
+
+## Credential File {#credential-file}
```

Comment:
Flagging this as a section that changed in @kevinlewi's PR, too.

Reply:
Yes, it was removed -- can you remove it here as well?